### PR TITLE
Propagate EKS connection secret updates

### LINKED ISSUE #35: Propagate EKS connection secret updates to bound claims
### What happened?

EKS clusters rotate their credentials frequently. The AWS stack currently does not propagate these updated credentials from the managed resource connection secret to its bound resource claim's connection secret. Full details in https://github.com/crossplaneio/crossplane-runtime/issues/35.

### How can we reproduce it?

See https://github.com/crossplaneio/crossplane-runtime/issues/35

### What environment did it happen in?
Crossplane version:

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes #35

EKS connection secrets change frequently. We must propagate them from the managed resource connection secret to the resource claim connection secret.

TODO: Test this manually to confirm it fixes the problem.

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the dependencies in [`app.yaml`] to include any new role permissions.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`app.yaml`]: https://github.com/crossplaneio/stack-aws/blob/master/config/stack/manifests/app.yaml

### CODE REVIEW

**Comment by negz** [comment]
> Bump module dependency?

@soorena776 is this a thing I need to do? My goal is to ensure this stack is built using the latest crossplane-runtime.

**Comment by soorena776** [comment]
@negz looks like https://github.com/crossplaneio/stack-aws/pull/20 is still not merged (it's ready to be merge by the way), so no go mod stuff needs to be done here.

**Comment by negz** [comment]
Turns out this needed some updates to the EKS controller to prevent it from overwriting its secret annotations each time it updated them. I've pushed those changes in a separate commit.

Everything appears to be in sync at T ~30 minutes from creation:

```
# Managed resource secret
$ kubectl -n aws-infra-dev get -o json ekscluster app-project1-dev-app-kubernetes-6mtlb |jq '.spec.writeConnectionSecretToRef.name'
"61cfb882-547c-4469-bc74-05f93b45af5b"

$ kubectl -n aws-infra-dev get -o json secret 61cfb882-547c-4469-bc74-05f93b45af5b| jq '.data'|md5
ca0f55dd5d8e7a5032527cf36d3f8e38

# Claim secret
$ diff -u <(kubectl -n aws-infra-dev get -o json secret 61cfb882-547c-4469-bc74-05f93b45af5b | jq .data) <(kubectl -n app-project1-dev get -o json secret k8scluster | jq .data)

# No difference between the two
$ kubectl -n app-project1-dev get -o json kubernetescluster app-kubernetes |jq '.spec.writeConnectionSecretToRef.name'
"k8scluster"
```

**Comment by negz** [comment]
```
$ kubectl -n app-project1-dev get -o json secret k8scluster|jq -r '.data.clusterCA|@base64d'> ca.crt
$ kubectl -s $(kubectl -n app-project1-dev get -o json secret k8scluster|jq -r '.data.endpoint|@base64d') --certificate-authority ca.crt --token $(kubectl -n app-project1-dev get -o json secret k8scluster|jq -r '.data.token|@base64d') get nodes
NAME                           STATUS   ROLES    AGE   VERSION
ip-172-31-41-28.ec2.internal   Ready    <none>   16m   v1.14.7-eks-1861c5
ip-172-31-43-80.ec2.internal   Ready    <none>   18m   v1.14.7-eks-1861c5
ip-172-31-82-51.ec2.internal   Ready    <none>   18m   v1.14.7-eks-1861c5
```

```
    Last Transition Time:  2019-10-17T02:03:36Z
    Reason:                Managed resource is available for use
    Status:                True
    Type:                  Ready
```

Claim connection secret is working ~20 minutes after the cluster became available. Will try again once an hour has passed. In the meantime @suskin @soorena776 PTAL - I pushed three more commits since you approved this PR.

**Comment by negz** [comment]
The claim secret is still working fine ~90 minutes after creation:

```
$ kubectl -s $(kubectl -n app-project1-dev get -o json secret k8scluster|jq -r '.data.endpoint|@base64d') --certificate-authority ca.crt --token $(kubectl -n app-project1-dev get -o json secret k8scluster|jq -r '.data.token|@base64d') get nodes
NAME                           STATUS   ROLES    AGE   VERSION
ip-172-31-41-28.ec2.internal   Ready    <none>   92m   v1.14.7-eks-1861c5
ip-172-31-43-80.ec2.internal   Ready    <none>   93m   v1.14.7-eks-1861c5
ip-172-31-82-51.ec2.internal   Ready    <none>   93m   v1.14.7-eks-1861c5
```

Sadly I accidentally hashed the endpoint, not the token, earlier so I'm not actually sure whether it has rotated yet or not. Here's the current working token:

```
$ kubectl -n app-project1-dev get -o json secret k8scluster|jq -r '.data.token|@base64d'|md5
d3732ed33884c9a83848619dcb64d006
```

File: pkg/controller/compute/eks.go [hunk 1]
```diff
@@ -108,10 +104,8 @@ type EKSClusterController struct{}
 // and Start it when the Manager is Started.
 func (c *EKSClusterController) SetupWithManager(mgr ctrl.Manager) error {
 	r := &Reconciler{
-		Client:     mgr.GetClient(),
-		scheme:     mgr.GetScheme(),
-		kubeclient: kubernetes.NewForConfigOrDie(mgr.GetConfig()),
-		recorder:   mgr.GetEventRecorderFor(controllerName),
+		Client:    mgr.GetClient(),
+		publisher: resource.NewAPISecretPublisher(mgr.GetClient(), mgr.GetScheme()),
```
Comment (soorena776) [inline pkg/controller/compute/eks.go]: nit: it seems odd that we are passing `mgr.GetClient()` twice to build a single object

File: pkg/controller/compute/eks.go [hunk 2]
```diff
@@ -133,22 +127,23 @@ func (r *Reconciler) fail(instance *awscomputev1alpha2.EKSCluster, err error) (r
...
+	if err := r.Get(ctx, n, s); err != nil {
 		return nil, err
 	}
 
-	// Connection Region must be with Spec.Region
-	if string(instance.Spec.Region) != config.Region {
-		config.Region = string(instance.Spec.Region)
+	// NOTE(negz): EKS clusters must specify a region for creation. They never
+	// use the provider's region. This should be addressed per the below issue.
+	// https://github.com/crossplaneio/stack-aws/issues/38
```
Comment (soorena776) [inline pkg/controller/compute/eks.go]: 👍

File: pkg/controller/compute/eks.go [hunk 3]
```diff
@@ -133,22 +127,23 @@ func (r *Reconciler) fail(instance *awscomputev1alpha2.EKSCluster, err error) (r
...
+	config, err := aws.LoadConfig(s.Data[p.Spec.Secret.Key], aws.DefaultSection, string(instance.Spec.Region))
```
Comment (soorena776) [inline pkg/controller/compute/eks.go]: FYI, there is also this method which does exactly that. 
https://github.com/crossplaneio/stack-aws/blob/master/pkg/controller/utils/utils.go#L19

File: pkg/controller/compute/eks.go [hunk 4]
```diff
@@ -61,10 +61,14 @@ const (
...
-	result        = reconcile.Result{}
-	resultRequeue = reconcile.Result{Requeue: true}
+	log = logging.Logger.WithName("controller." + controllerName)
+	ctx = context.Background()
+)
+
+// Amounts of time we wait before requeuing a reconcile.
+const (
+	aShortWait = 30 * time.Second
+	aLongWait  = 60 * time.Second
```
Comment (soorena776) [inline pkg/controller/compute/eks.go]: not a blocker for this pr, but I think these consts should be in the runtime library.

Comment (negz) [reply pkg/controller/compute/eks.go]: I see these as short term solutions, with the long term solution being using the managed resource reconciler to implement this controller.

https://godoc.org/github.com/crossplaneio/crossplane-runtime/pkg/resource#WithShortWait

I believe these were global constants for a while, until someone correctly pointed out that 30 and 60 seconds aren't necessarily the best requeue durations for all managed resource controllers. Instead now we make them the default, but allow controllers to override them if appropriate.

File: pkg/controller/compute/eks.go (hunk 3 above)
Comment (negz) [reply pkg/controller/compute/eks.go]: There's a slight difference in this one, in that we pass in the instance's region, not the provider's.

**Pull request merged**